clio-client/src/main/scala/org/broadinstitute/clio/client/webclient/ClioWebClient.scala

```diff
@@ -91,7 +70,26 @@ class ClioWebClient(
       maxQueuedRequests,
       OverflowStrategy.dropNew
     )
-    .via(connectionFlow)
+    .mapAsyncUnordered(maxConcurrentRequests) {
+      case (request, promise) => {
+
+        val responseStream = Source
+          .single(request)
+          .via(connectionFlow)
+          .completionTimeout(requestTimeout)
+
+        val response = responseStream
+          .recoverWithRetries(maxRequestRetries, {
+            case _ => responseStream
+          })
+          .runWith(Sink.head)
+
+        response.transformWith {
+          case Success(r)  => Future.successful(Success(r) -> promise)
+          case Failure(ex) => Future.successful(Failure(ex) -> promise)
```

Comment:
can you explain why this is returns `Future.successful()`?

Reply:
Yup! By default, `mapAsyncUnordered` will fail and shut down the web client's entire request-response stream if an element produces a failed `Future`, and we don't want that (otherwise apps like the backfill that use a single long-running client will fall over on any failure that pops up). There's a way to configure the stage to have it resume processing instead of starting, but in that case it'll still drop the failed element, which we also don't want.

What we do want is to have any `Failure`s preserved and passed to `promise.failure`, so the thread listening on the other end of the `promise` can receive the exception and handle it appropriately. We make this happen by transforming all the `Future[HttpResponse]`s here into successful `Future[Try[HttpResponse]]`s, so the final stage in the stream can handle both successes and failures and we never crash out the stream.

I dug a bit more into the streams API and I think there might be a cleaner way to do it. I'll try it out, and if I can't make it work I'll add a scaladoc explaining all this.

Reply:
Makes sense, thanks for the explanation.